src/include/catalog/postgres/pg_proc_impl.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "catalog/postgres/pg_proc.h"
+#include "common/managed_pointer.h"
+#include "execution/ast/builtins.h"
+#include "storage/projected_row.h"
+#include "storage/storage_defs.h"
+
+namespace noisepage::catalog {
+class DatabaseCatalog;
+}  // namespace noisepage::catalog
+
+namespace noisepage::execution::functions {
+class FunctionContext;
+}  // namespace noisepage::execution::functions
+
+namespace noisepage::storage {
+class RecoveryManager;
+class SqlTable;
+
+namespace index {
+class Index;
+}  // namespace index
+}  // namespace noisepage::storage
+
+namespace noisepage::transaction {
+class TransactionContext;
+}  // namespace noisepage::transaction
+
+namespace noisepage::catalog::postgres {
+class Builder;
+
+/** The NoisePage version of pg_proc. */
+class PgProcImpl {
+ public:
+  /**
+   * Prepare to create pg_proc.
+   * Does NOT create anything until the relevant bootstrap functions are called.
+   *
+   * @param db_oid          The OID of the database that pg_proc should be created in.
+   */
+  explicit PgProcImpl(db_oid_t db_oid);
```

Comment:
Specifically, you dislike that each pg_table has to maintain a db oid? An alternative is to pass in the db oid to all functions.

Reply:
Yeah, I realize that we need it to have the information for `StageWrite`, but intuitively this should be a transaction-level property since you cannot have a transaction cross database boundaries.  I guess I wonder if the alternative is to work this into the transaction context itself.

Reply:
The blast radius isn't that large if we wanted to refactor `StageWrite` and `TransactionContext`, at least not in the system proper -- tests/benchmarks yes. Will discuss in the meeting today if it makes sense.

Reply:
Will punt and write up an issue.

Reply:
Agree on the issue.  You can mark this resolved once its created.

Reply:
#1362